content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/advanced-design.md

```diff
@@ -0,0 +1,413 @@
+      return pointer;
+  }
+  
+  @NotNull
+  
+   public List<PublicKey> getParticipants() {
+      return List.of(...);
+  }
+}
+```
+
+## Designing Chainable Commands
+
+Chainable commands allows users to create, update, and delete chainable states.
+The `ChainableContractCreateCommand` creates new chainable states and will verify the following constraints:
+
+* On chainable state(s) creating, at least one chainable state must be created.
+* On chainable state(s) creating, the previous state pointer of every created chainable state must be null.
+
+```kotlin
+public final class Create extends ChainableContractCreateCommand<ExampleChainableState> {
+  
+  @NotNull
+  
+  public Class<ExampleChainableState> getContractStateType() {
+    return ExampleChainableState.class;
+  }
+  
+  @Override
+  
+   protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+    // Verify additional Create constraints
+  }
+}
+```
+
+The `ChainableContractUpdateCommand` supports updating existing chainable states and will verify the following constraints:
+
+* On chainable state(s) updating, at least one chainable state must be consumed.
+* On chainable state(s) updating, at least one chainable state must be created.
+* On chainable state(s) updating, the previous state pointer of every created chainable state must not be null.
+* On chainable state(s) updating, the previous state pointer of every created chainable state must be pointing to exactly one consumed chainable state, exclusively.
+
+```kotlin
+public final class Update extends ChainableContractUpdateCommand<ExampleChainableState> {
+  
+   @NotNull
+   
+    public Class<ExampleChainableState> getContractStateType() {
+     return ExampleChainableState.class;
+  }
+  
+   @Override
+  
+    protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+    
+    // Verify additional Update constraints
+  }
+}
+```
+
+The `ChainableContractDeleteCommand` supports deleting existing chainable states and will verify the following constraint:
+
+* On chainable state(s) deleting, at least one chainable state must be consumed.
+
+```kotlin
+
+    public final class Delete extends ChainableContractDeleteCommand<ExampleChainableState> {
+  
+    @NotNull
+    
+     public Class<ExampleChainableState> getContractStateType() {
+      return ExampleChainableState.class;
+  }
+  
+   @Override
+  
+    protected void onVerify(@NotNull final UtxoLedgerTransaction transaction) {
+    
+    // Verify additional Delete constraints
+    
+  }
+}
+```
+
+## Designing a Chainable Contract
+
+A chainable contract can be implemented by extending the `ChainableContract` class; for example:
+
+```kotlin
+public final class ExampleChainableContract extends ChainableContract {
+  
+   @Override
+  
+   public List<Class<? extends ChainableContractCommand<?>>> getPermittedCommandTypes() {
+      return List.of(Create.class, Update.class, Delete.class);
+      
+  }
+}
+```
```

Comment:
```suggestion
```kotlin
public final class ExampleChainableContract extends ChainableContract {
  @Override
  public List<Class<? extends ChainableContractCommand<?>>> getPermittedCommandTypes() {
    return List.of(Create.class, Update.class, Delete.class);  
  }
}
```
```